internal/cas/content.go

```diff
@@ -0,0 +1,224 @@
+package cas
+
+import (
+	"bufio"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/gruntwork-io/terragrunt/pkg/log"
+)
+
+// Content manages git object storage and linking
+type Content struct {
+	store *Store
+}
+
+const (
+	// DefaultDirPerms represents standard directory permissions (rwxr-xr-x)
+	DefaultDirPerms = os.FileMode(0755)
+	// StoredFilePerms represents read-only file permissions (r--r--r--)
+	StoredFilePerms = os.FileMode(0444)
+	// RegularFilePerms represents standard file permissions (rw-r--r--)
+	RegularFilePerms = os.FileMode(0644)
+)
+
+// NewContent creates a new Content instance
+func NewContent(store *Store) *Content {
+	return &Content{
+		store: store,
+	}
+}
+
+// Link creates a hard link from the store to the target path
+func (c *Content) Link(hash, targetPath string) error {
+	if err := c.ensureTargetDirectory(targetPath); err != nil {
+		return err
+	}
+
+	sourcePath := c.getPath(hash)
+
+	// Check if target exists
+	if _, err := os.Stat(targetPath); err == nil {
+		// File exists, skip creating link
+		return nil
+	} else if !os.IsNotExist(err) {
+		// Some other error occurred
+		return &WrappedError{
+			Op:   "stat_target",
+			Path: targetPath,
+			Err:  err,
+		}
+	}
+
+	// Create hard link
+	if err := os.Link(sourcePath, targetPath); err != nil {
+		// If hard link fails, try to copy the file
+		data, readErr := os.ReadFile(sourcePath)
+
+		if readErr != nil {
+			return &WrappedError{
+				Op:   "read_source",
+				Path: sourcePath,
+				Err:  ErrReadFile,
+			}
+		}
+
+		// Write to temporary file first
+		tempPath := targetPath + ".tmp"
+
+		if err := os.WriteFile(tempPath, data, RegularFilePerms); err != nil {
+			return &WrappedError{
+				Op:   "write_target",
+				Path: tempPath,
+				Err:  err,
+			}
+		}
+
+		// Rename to final path
+		if err := os.Rename(tempPath, targetPath); err != nil {
+			return &WrappedError{
+				Op:   "rename_target",
+				Path: tempPath,
+				Err:  err,
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *Content) ensureTargetDirectory(targetPath string) error {
+	targetDir := filepath.Dir(targetPath)
+	if err := os.MkdirAll(targetDir, DefaultDirPerms); err != nil {
+		return &WrappedError{
+			Op:   "create_target_dir",
+			Path: targetDir,
+			Err:  ErrCreateDir,
+		}
+	}
+
+	return nil
+}
+
+// Store stores a single content item
+func (c *Content) Store(l *log.Logger, hash string, data []byte) error {
+	c.store.mapLock.Lock()
+
+	if _, ok := c.store.locks[hash]; !ok {
+		c.store.locks[hash] = &sync.Mutex{}
+	}
+
+	c.store.locks[hash].Lock()
```

Comment:
It might work when only one Terragrunt process is running, but if two or more processes run in parallel (for example, on a CI/CD server), it could lead to conflicts when they both attempt to replace the same file.

Reply:
Added a modified timestamp based check to see if the temp file was created after the start of cloning, indicating a race condition.

There's still a chance that a race can occur, if one clone starts before another, and starts writing an entry before the other gets started, but that should be very rare. I'm doing the timestamp based comparison because any failed clone might leave behind temporary files that need to be ignored.